src/js/BookReader/BookModel.js

```diff
@@ -0,0 +1,286 @@
+   * null if there is no facing page or the index is invalid.
+   * @param  {number} pindex
+   * @return {array} - eg [0, 1]
+   */
+  getSpreadIndices(pindex) {
+    if (this.br.pageProgression == 'rl') {
+      return this.getPageSide(pindex) == 'R' ? [pindex + 1, pindex] : [pindex, pindex - 1];
+    } else {
+      return this.getPageSide(pindex) == 'L' ? [pindex, pindex + 1] : [pindex - 1, pindex];
+    }
+  }
+
+  /**
+   * Single images in the Internet Archive scandata.xml metadata are (somewhat incorrectly)
+   * given a "leaf" number.  Some of these images from the scanning process should not
+   * be displayed in the BookReader (for example colour calibration cards).  Since some
+   * of the scanned images will not be displayed in the BookReader (those marked with
+   * addToAccessFormats false in the scandata.xml) leaf numbers and BookReader page
+   * indexes are generally not the same.  This function returns the BookReader page
+   * index given a scanned leaf number.
+   *
+   * This function is used, for example, to map between search results (that use the
+   * leaf numbers) and the displayed pages in the BookReader.
+   * @param {number} leafNum
+   * @return {number}
+   */
+  leafNumToIndex(leafNum) {
+    const index = this._getDataFlattened()
+      .findIndex(d => d.leafNum == leafNum);
+    // If no match is found, fall back to the leafNum provide (leafNum == index)
+    return index > -1 ? index : leafNum;
+  }
+
+  /**
+   * Parses the pageString format
+   * @param {string} pageNum
+   * @return {number|undefined}
+   */
+  parsePageString(pageNum) {
+    let pageIndex;
+    // Check for special "leaf"
+    const leafMatch = /^leaf(\d+)/.exec(pageNum);
+    if (leafMatch) {
+      pageIndex = this.leafNumToIndex(parseInt(leafMatch[1], 10));
+      if (pageIndex === null) {
+        pageIndex = undefined; // to match return type of getPageIndex
+      }
+    } else {
+      pageIndex = this.getPageIndex(pageNum);
+    }
+    return pageIndex;
+  }
+
+  /**
+   * Helper. Flatten the nested structure (make 1d array),
+   * and also add pageSide prop
+   * @return {PageData[]}
+   */
+  _getDataFlattened() {
+    if (this._getDataFlattenedCached && this._getDataFlattenedCached[1] === this.br.data.length)
+      return this._getDataFlattenedCached[0];
+
+    let prevPageSide = null;
+    // @ts-ignore TS doesn't know about flatMap for some reason
+    const flattend = this.br.data.flatMap(spread => {
+      return spread.map(page => {
+        if (!page.pageSide) {
+          if (prevPageSide === null) {
+            page.pageSide = spread.length === 2 ? 'L' : 'R';
+          } else {
+            page.pageSide = prevPageSide === 'L' ? 'R' : 'L';
+          }
+        }
+        prevPageSide = page.pageSide;
+        return page;
+      });
+    });
+
+    // length is used as a cache breaker
+    this._getDataFlattenedCached = [flattend, this.br.data.length];
```

Comment:
Ideally this might be implemented using a decorator; i.e.

```js
@memoize
_getDataFlattenedCached() { ... }
```

Then it would essentially be stateless (although the state would live somewhere else). Then it would be state-less (relative to the class it's a part of). But I'm happy to drop the "Helper" word atm :+1: